posix/device.c

```diff
@@ -0,0 +1,321 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#include <openenclave/enclave.h>
+
+#include <openenclave/bits/safecrt.h>
+#include <openenclave/corelibc/errno.h>
+#include <openenclave/corelibc/stdio.h>
+#include <openenclave/corelibc/stdlib.h>
+#include <openenclave/corelibc/string.h>
+#include <openenclave/internal/posix/device.h>
+#include <openenclave/internal/posix/raise.h>
+#include <openenclave/internal/raise.h>
+#include <openenclave/internal/thread.h>
+#include <openenclave/internal/trace.h>
+#include <openenclave/internal/utils.h>
+
+/*
+**==============================================================================
+**
+** Define the table of file-descriptor entries:
+**
+**==============================================================================
+*/
+
+#define TABLE_CHUNK_SIZE 64
+
+typedef oe_device_t* entry_t;
+static entry_t* _table;
+static size_t _table_size;
+static oe_spinlock_t _lock = OE_SPINLOCK_INITIALIZER;
+static bool _installed_atexit_handler;
+
+static void _atexit_handler(void)
+{
+    oe_free(_table);
+}
+
+static int _resize_table(size_t new_size)
+{
+    int ret = -1;
+
+    if (!_installed_atexit_handler)
+    {
+        oe_atexit(_atexit_handler);
+        _installed_atexit_handler = true;
+    }
+
+    /* Round the new capacity up to the next multiple of the chunk size. */
+    new_size = oe_round_up_to_multiple(new_size, TABLE_CHUNK_SIZE);
+
+    if (new_size > _table_size)
+    {
+        entry_t* p;
+        size_t n = new_size;
+
+        /* Reallocate the table. */
+        if (!(p = oe_realloc(_table, n * sizeof(entry_t))))
+            goto done;
+
+        /* Zero-fill the unused portion. */
+        {
+            const size_t num_bytes = (n - _table_size) * sizeof(entry_t);
+
+            if (oe_memset_s(p + _table_size, num_bytes, 0, num_bytes) != OE_OK)
+                goto done;
+        }
+
+        _table = p;
+        _table_size = new_size;
+    }
+
+    ret = 0;
+
+done:
+    return ret;
+}
+
+/* Raise and log an error if the condition is false. */
+#define CHECK_CONDITION(COND)                                         \
+    do                                                                \
+    {                                                                 \
+        if (!(COND))                                                  \
+            OE_RAISE_ERRNO_MSG(OE_EINVAL, "failed check: %s", #COND); \
+    } while (0)
+
+static int _check_device(oe_device_t* device)
```

Comment:
Shouldn't this method return -1 on an unrecognized type? (i.e. needs a `default: goto done;` case?)

Reply:
I added a default case, but strictly speaking it is not needed since the compiler will warn if there is a new enum type that is not covered.